numpy/ma/core.py

```diff
@@ -2978,9 +2976,19 @@ def __array_finalize__(self, obj):
             # heuristic it's not bad.) In all other cases, we make a copy of
             # the mask, so that future modifications to 'self' do not end up
             # side-effecting 'obj' as well.
-            if (obj.__array_interface__["data"][0]
+            if (_mask is not nomask and obj.__array_interface__["data"][0]
                     != self.__array_interface__["data"][0]):
-                _mask = _mask.copy()
+                # We should make a copy. But we could get here via astype,
+                # in which case the mask might need a new dtype as well
+                # (e.g., changing to or from a structured dtype), and the
+                # order could have changed. Cover these with astype if the
+                # dtype is not the same (as it would be in slices).
+                order = "F" if self.flags.f_contiguous else "C"
+                if self.dtype == obj.dtype:
```

Comment:
I considered using `astype` unconditionally, but that adds 2-3 micro-sec, which I felt was a bit much for things like `copy` (I realize now that I write this that slices are actually not a problem here, since nominally we know the data are copied.

Reply:
>  but that adds 2-3 micro-sec

Are you sure that time doesn't come from `make_mask_descr`? I'd be inclined to use `astype`, but only compute a new mask type if the dtype changed.